### ISSUE TITLE
Samples of segmentation-guided ddpm/ddim NOT consistent with their respective segmentation masks during training/inference (cardiac MRI)

### ISSUE THREAD

**Issue by varshak97** [mazurowski-lab/segmentation-guided-diffusion#17]
Hello!
I want to begin by expressing my appreciation for this work; particularly the straightforward mask-ablated training strategy. Thank you to you and your team for sharing the code!

I am facing an issue when training the segmentation-guided DDPM/DDIM model with cardiac MRI data (using the public ACDC dataset). I have observed that the generated samples are not consistent with their corresponding segmentation masks, during training/inference.

**Details:**
Inconsistent Results: The reconstructed/generated samples do not align with the input segmentation masks, i.e. the spatial features outlined by the masks are not correctly reflected in the generated samples.
Noisy Samples: There is a noticeable presence of noisy samples, although increasing the number of epochs has somewhat resolved this issue. After training for 1000 epochs, there are still occasional noisy outputs, but my primary concern is the inconsistency between the generated samples and segmentation masks.

I have double-checked my data preparation step, and I don't believe the issue lies there. This is command I used for training the segmentation-guided diffusion model:

>  python3 segmentation-guided-diffusion/main.py \
               --mode "train" \
               --img_size 256 \
               --num_img_channels 1 \
               --dataset "cardiac_mri" \
               --img_dir "/home/varshak/scratch/acdc/segdiff/data_dir/" \
               --seg_dir "/home/varshak/scratch/acdc/segdiff/mask_dir/" \
               --model_type "DDPM" \
               --segmentation_guided \
               --num_segmentation_classes 4 \
               --train_batch_size 16 \
               --eval_batch_size 8 \
               --use_ablated_segmentations \
               --eval_noshuffle_dataloader  \
               --num_epochs 1000

Following are sample outputs from epoch 999 for more context. Also, this issue persists throughout the entire training process, not just at the beginning or the end.
I would appreciate any insights or suggestions on how to resolve this issue! Thank you so much for your time :)

Results/samples of ddpm_cardiac-mri_256_segguided_abalated @ Epoch 999:

0999_cond_seg_all.png
![0999_cond_seg_all](https://github.com/user-attachments/assets/6f11705c-74e5-4ed0-94d3-7d76f0a4d182)

0999_orig.png
![0999_orig](https://github.com/user-attachments/assets/3c5ef9c9-de8b-4ed9-a703-29f7a0b9e484)

0999.png
![0999](https://github.com/user-attachments/assets/a4d0ca33-e969-4f38-918e-289126718df3)

**Comment by varshak97** [comment]
Hi, I'm attaching a drive link to a Google Doc with additional samples (training and inference) just for further clarification and review regarding this issue.

Link: https://docs.google.com/document/d/17VbNJ9LZJPoskOSP2kckq5vs9L7E2h6X-Yb-iZVUzFw/edit?usp=sharing

**Comment by nickk124** [comment]
Hi, thanks for trying out our code!

I think that the reason the model doesn't always follow your segmentation prompts is due to using mask ablated training, combined with the small size of the objects. It seems like the model sometimes follows the prompts, but not always. I think that this could be due to a number of factors:
1. by default, the model will see each of the 2^(C-1) = 8 possibilities of classes removed from the input mask in training with equal probability. So it may be stretched to thin in learning to follow all possibilities of these conditioning types at once.
2. the small size of the objects provided fairly weak guidance for the model, so it may not always follow the prompts that well, especially due to the variability created by the MAT.

This is definitely a limitation of the MAT algorithm which would be assisted by future work, but for now, I'd suggest deciding which types of segmentation conditions you want the model to follow the best, in terms of what classes are present vs. missing in the masks, and then customize the probability of removing/keeping certain classes from the masks during training, which can be done by adding a custom option to the if/elif tree in the function `ablate_masks()` at https://github.com/mazurowski-lab/segmentation-guided-diffusion/blob/9a31f0f4eef0d7b835f01b2a6301df854b8040e9/eval.py#L324. Does that help?

As for the noisy samples, I'm glad that more epochs worked, but I would also try a lower learning rate. You could also try tuning the noise scheduler, but I doubt that that is necessary.

**Comment by varshak97** [comment]
I agree that using the MAT algorithm as it is in the repository may not be suitable for this cardiac MRI dataset, given the size of the objects. Earlier this week, I experimented with training without the MAT algorithm by removing the `use_ablated_segmentations` argument, and the generated images aligned quite well with the segmentations! I plan to customize the `ablate_masks()` function to better suit my requirements and see how the results vary. Thank you for the suggestion!  

Regarding the noisy samples, I tested two learning rates, 2e-5 and 1e-4. For SegDiff-MAT, I found that a learning rate of 2e-5 combined with training for more epochs produced significantly better results than 1e-4. However, for SegDiff-STD, a lower learning rate with more epochs (~900 epochs) led to overfitting i.e. some generated samples appeared memorised. In this case, a learning rate of 1e-4 for 600 epochs was a better choice.  
In conclusion, some hyperparameter tuning is necessary to address the noisy samples issue, but it seems manageable. I'll also try tuning the noise scheduler to see if it makes any difference. Thanks again!

**Comment by nickk124** [comment]
Glad to hear that some testing and tuning worked out! I'm also curious if your dataset is small, which may be a cause of some of these issues (e.g., overfitting). Anyways, I'll close this for now because your initial questions seem resolved, but I'm happy to re-open if anything comes up!
